### PR TITLE
Non-record: Diffusion-Noised Teacher AR Hybrid (val_bpb=1.2734, 8xH100)

### PR DESCRIPTION
**Pull request by anthony-maio**
## Non-Record Submission: Diffusion-Noised Teacher Forcing

Answers OpenAI's explicit request for **text diffusion** submissions.

### Approach
Standard AR transformer with a diffusion-inspired auxiliary loss: during training, a fraction of input tokens are corrupted (masked or randomly replaced), and the model trains on both clean and noisy forward passes with interpolated loss:
```
loss = lerp(clean_loss, noisy_loss, 0.35)
```
Noise ratio ramps from 5% to 35% over training, forcing the model to learn robust representations.

### Results (8xH100 SXM, 600s, seed=1337)
- **val_bpb: 1.2734** (int8+zlib roundtrip)
- val_loss: 2.1500
- Steps: 3,227 at 186ms/step
- Training time: 600s (wallclock cap)
- Artifact: 15.8MB (under 16MB)
- Baseline 9L architecture (no SmearGate/BigramHash/XSA/FA3)

### Key Observations
- Diffusion auxiliary loss trains stably with no instability
- Clean loss (2.15) consistently lower than noisy loss (2.50) — model learns to distinguish
- Gap widens as noise ratio increases — expected behavior
- Not competitive with SOTA but demonstrates the concept works

### Included files
- `train_gpt.py` — training script
- `train_8xH100.log` — full 8xH100 training log (seed=1337)
- `train.log` — local smoke test log (4-step, 1xGPU)
- `submission.json`
- `README.md`

### Credits
- Baseline architecture: OpenAI Parameter Golf starter code
- Diffusion concept: inspired by MDLM, SEDD discrete diffusion literature

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Diffusion-Noised Teacher AR Hybrid (val_bpb=1.2734, 8xH100)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

--- ## Analysis ### N-gram family bug (ILLEGAL: target XOR'd into hash key, NOT BigramHash) NOT PRESENT. No BigramHash, no n-gram hash structures, no XOR operations on target tokens anywhere in the file. Clean. ### Pre-Quant TTT (ILLEGAL: multi-epoch on val_tokens without score-first) NOT PRESENT. The TTT evaluation (`eval_val_ttt_lora`) is invoked at line 1467 only if `ttt_eval_enabled` is set, and it is called after the quantization roundtrip (lines 1421-1464). The base model at that point has already been quantized and reloaded. There is no multi-epoch loop over val_tokens; each document batch is processed once with one Adam step per chunk. ### Score-first TTT (LEGAL: PR #1413 pattern, is_last_chunk guard) PRESENT AND LEGAL. In `eval_val_ttt_lora` (lines 904-1011), the per-chunk loop at lines 956-1002 follows the correct causal order: - Line 986-994: **Score first** — `_accumulate_bpb` accumulates loss and byte counts for BPB before any weight update. Comment explicitly states "before training on chunk". - Lines 997-1002: **Train after** — Adam step on LoRA params only when `needs_train` is True (line 961: `any(ci < nc - 1 for nc in num_chunks)`). The `needs_train` guard is equivalent to the `is_last_chunk` guard in PR #1413: training is skipped for the final chunk of each document (when `ci == nc - 1`), so the last chunk's contribution to the score is computed without a preceding update on that chunk. ### Scored-region SLOT NOT PRESENT. No external data, scored-region caches, or lookup tables beyond the standard tokenizer byte-count LUTs used for BPB calculation. ### Architecture Standard...

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
